src/core/server/saved_objects/service/lib/repository.ts

```diff
@@ -1515,26 +1515,62 @@ export class SavedObjectsRepository {
   }
 
   /**
-   * Increases a counter field by one. Creates the document if one doesn't exist for the given id.
+   * Increments all the specified counter fields by one. Creates the document
+   * if one doesn't exist for the given id.
+   *
+   * @remarks
+   * When using incrementCounter for collecting usage data, you need to ensure
+   * that usage collection happens on a best-effort basis and doesn't
+   * negatively affect your plugin or users (see the example):
+   *  - Swallow any exceptions thrown from the incrementCounter method and log
+   *    a message in development.
+   *  - Don't block your application on the incrementCounter method (e.g.
+   *    don't use `await`)
+   * @remarks
+   * When supplying a field name like `stats.api.counter` the field name will
+   * be used as-is to create a document like:
+   *   `{attributes: {'stats.api.counter': 1}}`
+   * It will not create a nested structure like:
+   *   `{attributes: {stats: {api: {counter: 1}}}}`
+   *
+   * @example
+   * Collecting usage data
+   * ```ts
+   * const repository = coreStart.savedObjects.createInternalRepository();
+   *
+   * // NOTE: Usage collection happens on a best-effort basis, so we don't
+   * // `await` the promise returned by `incrementCounter` and we swallow any
+   * // exceptions in production.
+   * repository
+   *   .incrementCounter('test_counter_type', 'counter_2', [
+   *     'stats.api.count',
+   *     'stats.api.count2',
+   *     'stats.total',
+   *   ])
+   *   .catch((e) => (coreContext.env.cliArgs.dev ? logger.error(e) : e));
+   * ```
    *
    * @param {string} type
    * @param {string} id
-   * @param {string} counterFieldName
+   * @param {string} counterFieldNames
    * @param {object} [options={}]
    * @property {object} [options.migrationVersion=undefined]
    * @returns {promise}
    */
   async incrementCounter(
     type: string,
     id: string,
-    counterFieldName: string,
+    counterFieldNames: string[],
```

Comment:
Why are we introducing a breaking change in the API instead of allowing `counterFieldNames: string | string[]`? Do we really want to get rid of the plain `string` signature that badly?

Reply:
It's not really a big deal in this specific case (we could go either way without a big cost).

In general I dislike API's that aim to be convenient by allowing you to pass in many shapes of data. This makes it harder to understand the consuming code (what does it mean when it's a string vs array?) but also introduces complexity in the implementation (now we have to test both code paths, not relevant in this case, but if there's an HTTP API or other layers it also means testing both code paths in all the layers).

So in this specific case it would be simple and easy enough to not break the API, but it also felt like breaking the API is quite safe and keeps the code just a little bit simpler.